### PR TITLE
[onert] Update Categorical crossentropy implementation

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit updates Categorical crossentropy implementation to apply changed output shape. The output shape is changed from [1] to [batch_size].

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

### CODE REVIEW

**Comment by hseok-oh** [comment]
```
[2023-11-30T12:38:40.330Z] [  FAILED  ] 2 tests, listed below:
[2023-11-30T12:38:40.330Z] [  FAILED  ] CKer_Operation.LossCategoricalCrossEntropy
[2023-11-30T12:38:40.330Z] [  FAILED  ] CKer_Operation.LossCategoricalCrossEntropyGrad
[2023-11-30T12:38:40.330Z] 
[2023-11-30T12:38:40.330Z]  2 FAILED TESTS
[2023-11-30T12:38:40.330Z] Product/out/unittest/test_cker failed... return code: 1
```

**Comment by jyoungyun** [comment]
Well,,, my local machine passed these test cases. However, some environments make an error in th
is calculation. The `EXPECT_NEAR` is one of solutions. I will check the appropriate `abs_error` value.

```
[2023-11-30T12:38:40.042Z] [ RUN      ] CKer_Operation.LossCategoricalCrossEntropy
[2023-11-30T12:38:40.042Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:51: Failure
[2023-11-30T12:38:40.042Z] Expected equality of these values:
[2023-11-30T12:38:40.042Z]   output[i]
[2023-11-30T12:38:40.042Z]     Which is: 0.046494164
[2023-11-30T12:38:40.042Z]   expected[i]
[2023-11-30T12:38:40.042Z]     Which is: 0.046494246
[2023-11-30T12:38:40.042Z] [  FAILED  ] CKer_Operation.LossCategoricalCrossEntropy (0 ms)
[2023-11-30T12:38:40.042Z] [ RUN      ] CKer_Operation.neg_LossCategoricalCrossEntropy
[2023-11-30T12:38:40.042Z] [       OK ] CKer_Operation.neg_LossCategoricalCrossEntropy (0 ms)
[2023-11-30T12:38:40.042Z] [ RUN      ] CKer_Operation.LossCategoricalCrossEntropyGrad
[2023-11-30T12:38:40.042Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:85: Failure
[2023-11-30T12:38:40.042Z] Expected equality of these values:
[2023-11-30T12:38:40.042Z]   output[i]
[2023-11-30T12:38:40.043Z]     Which is: -672.35242
[2023-11-30T12:38:40.043Z]   expected[i]
[2023-11-30T12:38:40.043Z]     Which is: -672.35638
[2023-11-30T12:38:40.043Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:85: Failure
[2023-11-30T12:38:40.043Z] Expected equality of these values:
[2023-11-30T12:38:40.043Z]   output[i]
[2023-11-30T12:38:40.043Z]     Which is: -2.0280554
[2023-11-30T12:38:40.043Z]   expected[i]
[2023-11-30T12:38:40.043Z]     Which is: -2.0280604
[2023-11-30T12:38:40.043Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:85: Failure
[2023-11-30T12:38:40.043Z] Expected equality of these values:
[2023-11-30T12:38:40.043Z]   output[i]
[2023-11-30T12:38:40.043Z]     Which is: -1.0926754
[2023-11-30T12:38:40.043Z]   expected[i]
[2023-11-30T12:38:40.043Z]     Which is: -1.0926765
[2023-11-30T12:38:40.043Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:85: Failure
[2023-11-30T12:38:40.043Z] Expected equality of these values:
[2023-11-30T12:38:40.043Z]   output[i]
[2023-11-30T12:38:40.043Z]     Which is: -2371.3101
[2023-11-30T12:38:40.043Z]   expected[i]
[2023-11-30T12:38:40.043Z]     Which is: -2371.3149
```

**Comment by jyoungyun** [comment]
> Is it difficult to modify the test values so that expected values are smaller?

Yes, I can. However, I'm looking for a reason why the error is bigger than before. 👀

**Comment by ragmani** [comment]
> However, I'm looking for a reason why the error is bigger than before. 

I don't know the exact reason why there are floating point errors. It may related to difference of SIMD, architectures, optimizations or others.
According to https://stackoverflow.com/questions/75535438/increased-floating-point-error-when-using-simd-intrinsics-or-maybe-even-the-oth, "FP math is not associative(결합법칙); optimizing as if it is will change the results.", and "Unless you were taking special numerical precautions like adding up the small numbers first, the sequential-order result isn't "more correct", they just have different errors."

**Comment by hseok-oh** [comment]
```
[2023-12-06T02:52:36.694Z] [  FAILED  ] 1 test, listed below:
[2023-12-06T02:52:36.694Z] [  FAILED  ] CKer_Operation.LossCategoricalCrossEntropyGrad
[2023-12-06T02:52:36.694Z] 
[2023-12-06T02:52:36.694Z]  1 FAILED TEST
[2023-12-06T02:52:36.694Z] Product/out/unittest/test_cker failed... return code: 1
```

**Comment by hseok-oh** [comment]
`CKer_Operation.LossCategoricalCrossEntropyGrad` is passed on my environment (ubuntu 22.04)

**Comment by jyoungyun** [comment]
```
[2023-12-06T09:49:12.100Z] /opt/jenkins_agent/workspace/nnfw/master/pr-onert-cross-debug/compute/cker/src/train/Loss.test.cc:85: Failure
[2023-12-06T09:49:12.100Z] The difference between output[i] and expected[i] is 34359738368, where
[2023-12-06T09:49:12.100Z] output[i] evaluates to -1.6051361373015245e+17,
[2023-12-06T09:49:12.100Z] expected[i] evaluates to -1.6051364808989082e+17.
[2023-12-06T09:49:12.100Z] The abs_error parameter 1e-3f evaluates to 0.0010000000474974513 which is smaller than the minimum distance between doubles for numbers of this magnitude which is 32, thus making this EXPECT_NEAR check equivalent to EXPECT_EQUAL. Consider using EXPECT_DOUBLE_EQ instead.
```

**Review by ragmani — APPROVED** [review]
LGTM

**Comment by jyoungyun** [comment]
Finally this pr is ready to review. Please take another look. :)

**Comment by jyoungyun** [comment]
This PR should be merged to support changed loss output shape. Related PR is #12135.

**Review by hseok-oh — APPROVED** [review]
LGTM
